bigcommerce/grphp
# ARCH-532: grphp 1.0

## What? Why?

grphp 1.0. Features:

* Lazily load client on call instead of client instantiation to delay channel creation as late as possible
* Support for PHP 7.0+ only, add `strict_types` to all files
* Move authentication into interceptors to allow control of authentication timings
* Simplify configuration and add type checking
* Add channel configuration support
* Add interceptor registry for future insertion order support
* Add PSR2 code linter to test suite

## How to test

Unit tests have pretty good test coverage. Also added a script here: https://gist.github.com/splittingred/710f9ef28124e4b57de22d2c492ec96e - you'll need to replace paths/store ID, and make sure storeconfig is running in CDVM, but this will do a full e2e call. You can adjust store ID to simulate errors too.

## A Few Other Notes

There's a few breaking changes here (that align us better with the native gRPC implementation while still providing an anti-corruption layer to shield us from upstream changes:

### Interceptors

You now add interceptors to a registry, like so:

```php
$client->interceptors->add(new MyInterceptor($options));
```

They execute in the order they are added to the registry.

### Auth

Authentication is now an interceptor just like any other:

```php
$client->interceptors->add(new \Grphp\Client\Interceptors\Authentication\Basic([
    'username' => 'my-name',
    'password' => 'my-pass'
]));
```

The config object is now type-safe and does not support an untyped array in initializer; you must set via accessors.

```php
$config = new Grphp\Client\Config();
$config->setHostname('linkerd:4142');
$config->setErrorSerializer(new \Bigcommerce\Interfaces\Serializers\Error());
$client = new Grphp\Client(\Bigcommerce\Rpc\Storeconfig\StoreConfigClient::class, $config);
```

----

@bigcommerce/platform-engineering @precariouspanther @PascalZajac @jmwiese @aleachjr @filakhtov
